### PR TITLE
[WIP] Core/Player: Titan's Grip should only allow dual wielding 2H Axes, Maces and Swords

### LINKED ISSUE #20972: Titans Grip Allows For Dual Fishing Pole Buffs
**Description:**

The warrior class talent Titans Grip with the Dual Wield spell and fishing skill allows for equipping two fishing poles. The tool-tip reads Axes, Maces, Swords and the buff allows for polearms and staves. Fishing poles are not true melee weapons and shouldn't be allowed to dual wield.

The side effect of this is that additional fishing skill buffs stack on the second weapon and its lure.

**Current behaviour:**

Equipping two fishing poles is possible with TG and DW and Fishing Skill.

**Expected behaviour:**

Not allowed to equip a second pole, similar to if TG was not talented. Only "weapons" can be dual wielded.

**Steps to reproduce the problem:**

1. Learn TG, DW, Fishing
2. Equip two Strong Poles
3. View fishing skill is 1+10=11

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

TrinityCore rev. eba1c1ff97af 2017-11-25 11:21:12 +0100 (3.3.5 branch) (Unix, RelWithDebInfo, Static) (worldserver-daemon)

**TDB version:**  TDB_full_world_335.63_2017_04_18.sql

**Operating system:** Linux

### PR DESCRIPTION
**Pull request by svranesevic**
**Changes proposed:**

-  Disallow equipping fishing pole in offhand

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #20972

**Tests performed:**
- [x] Builds
- [x] Tested in-game

### CODE REVIEW

**Comment by Ascathor** [comment]
Are there more types of weapon that shouldn't be allowed ?

**Comment by svranesevic** [comment]
@Ascathor Don't think so, this addresses fishing poles issue with Titan's Grip.
Open to all suggestion tho.

**Comment by gmillerd** [comment]
Thanks, 20972.

I believe the issue is that Titans Grip should only work with melee weapons. A fishing pole is an equipable item that does damage (like a weapon), but isn't one.

There could be quest items that should not be equipped as such (put quest item in offhand, keep melee weapon in main hand).

**Comment by svranesevic** [comment]
@gmillerd Should we just close this PR then, and expand #20972 with all problems concerning Titan's Grip dual wielding ?

**Comment by gmillerd** [comment]
@svranesevic I don't know the TC process and would defer, but your fix works.

After seeing your fix it seems ItemSubclassWeapon, locking down TG to only use weapons seems pretty easy. Thanks

**Comment by tkrokli** [comment]
Seeing as this PR will close a specific issue dealing with a specific item, I would agree to this solution and approve of a merge. This fits with the general idea of making smaller PRs with simple solutions, unless one or more of the main TC developers disagree with the solution, requesting changes to create a more general solution. We must wait patiently for more feedback on this.

**Comment by Ascathor** [comment]
@tkrokli I would actually disagree with the fix. If there are more types of weaponry that are not supposed to be duo-wield by titan grip, we need a more generic, less-hacky fix. 

This fix targets a type of weapon. However, as it was previously stated, there are possibly more cases than fishing rods. So I would propose more investigation and checking if there are additional cases that have to be covered.

**Comment by ariel-** [comment]
Equipable items that are not weapons? These two came to my mind:

http://www.wowhead.com/item=10515
http://www.wowhead.com/item=41755

**Comment by svranesevic** [comment]
Alright, I'll get onto this.
Once again, should we close this PR, or we can mark it WIP and I'll update it along ?

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -9548,7 +9548,18 @@ uint8 Player::FindEquipSlot(ItemTemplate const* proto, uint32 slot, bool swap) c
             // suggest offhand slot only if know dual wielding
             // (this will be replace mainhand weapon at auto equip instead unwonted "you don't known dual wielding" ...
             if (CanDualWield())
-                slots[1] = EQUIPMENT_SLOT_OFFHAND;
+            {
+                if (CanTitanGrip())
+                {
+                    Item* mainItem = GetItemByPos(INVENTORY_SLOT_BAG_0, EQUIPMENT_SLOT_MAINHAND);
+                    if (mainItem && mainItem->GetTemplate()->CanBeDualWieldedWithTitansGrip() && proto->CanBeDualWieldedWithTitansGrip())
+                    {
```
Comment (Keader) [inline src/server/game/Entities/Player/Player.cpp]: brackets not needed

File: src/server/game/Entities/Item/ItemTemplate.cpp [hunk 2]
```diff
@@ -129,6 +129,34 @@ uint32 ItemTemplate::GetSkill() const
...
+
+    return SubClass == subClass;
+}
+
+bool ItemTemplate::IsWeaponOfSubClass(std::initializer_list<ItemSubclassWeapon> subClasses) const
+{
+    if (Class != ITEM_CLASS_WEAPON)
+        return false;
+
+    for (const auto& subClass : subClasses)
```
Comment (ariel-) [inline src/server/game/Entities/Item/ItemTemplate.cpp]: auto const&

File: src/server/game/Entities/Item/ItemTemplate.cpp [hunk 3]
```diff
@@ -129,6 +129,34 @@ uint32 ItemTemplate::GetSkill() const
...
+        if (IsWeaponOfSubClass(subClass))
+            return true;
+
+    return false;
+}
+
+bool ItemTemplate::CanBeDualWieldedWithTitansGrip() const
+{
```
Comment (ariel-) [inline src/server/game/Entities/Item/ItemTemplate.cpp]: Titan's Grip already has SubclassMask defined, maybe use that?

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -11340,7 +11330,35 @@ InventoryResult Player::CanEquipItem(uint8 slot, uint16 &dest, Item* pItem, bool
...
                         return swap ? EQUIP_ERR_ITEMS_CANT_BE_SWAPPED : EQUIP_ERR_INVENTORY_FULL;
                 }
+                else
+                {
+                    if (eslot != EQUIPMENT_SLOT_MAINHAND && eslot != EQUIPMENT_SLOT_OFFHAND)
+                        return EQUIP_ERR_ITEM_CANT_BE_EQUIPPED;
+
+                    if (eslot == EQUIPMENT_SLOT_MAINHAND)
+                    {
+                        if (Item* offItem = GetItemByPos(INVENTORY_SLOT_BAG_0, EQUIPMENT_SLOT_OFFHAND))
```
Comment (ariel-) [inline src/server/game/Entities/Player/Player.cpp]: bracket here

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (tkrokli) [reply src/server/game/Entities/Player/Player.cpp]: Yes, when there is only one line below or the completion of the line above (making it the same line), it is also easier to read the code without the redundant curly brackets (line 9556 + 9558) (in addition to being redundant codestyle-wise).

File: src/server/game/Entities/Item/ItemTemplate.cpp (hunk 3 above)
Comment (svranesevic) [reply src/server/game/Entities/Item/ItemTemplate.cpp]: Not sure what you think here, little help ?

Comment (ariel-) [reply src/server/game/Entities/Item/ItemTemplate.cpp]: Titan's Grip spell:

```
ID - 46917 Titan's Grip (Rank 1)
=================================================
Description: Allows you to equip two-handed axes, maces and swords in one hand.  While you have a two-handed weapon equipped in one hand, your physical damage done is reduced by $49152s1%.
Category = 0, SpellIconID = 2750, activeIconID = 0, SpellVisual = (0,0)
Family SPELLFAMILY_WARRIOR, flag [0] 0x00000000 [1] 0x00000000 [2] 0x00000000

SpellSchoolMask = 1 (SPELL_SCHOOL_MASK_NORMAL)
DamageClass = 0 (SPELL_DAMAGE_CLASS_NONE)
PreventionType = 0 (SPELL_PREVENTION_TYPE_NONE)
=================================================
Attributes: 0x000001C0 (SPELL_ATTR0_PASSIVE, SPELL_ATTR0_HIDDEN_CLIENTSIDE, SPELL_ATTR0_HIDE_IN_COMBAT_LOG)
=================================================
Skill (Id 256) "Fury"
    ReqSkillValue 1, Forward Spell = 0, MinMaxValue (0, 0), CharacterPoints (0, 0)Spell Level = 0, base 0, max 0, maxTarget 0
EquippedItemClass = 2 (WEAPON)
    SubClass mask 0x0000A1B3 (AXE, AXE2, MACE, MACE2, SWORD, SWORD2, FIST, DAGGER)
    InventoryType mask = 0x00600000 (WEAPONMAINHAND, WEAPONOFFHAND)
```

Comment (svranesevic) [reply src/server/game/Entities/Item/ItemTemplate.cpp]: I see now, thank you

**Comment by svranesevic** [comment]
@Keader Did not manage to pick it up since, will try to wrap it up over the weekend, or anyone can pick it up from here as functionality is implemented and remaining tasks are to format code according to TC code style and https://github.com/TrinityCore/TrinityCore/pull/21145#discussion_r159680166

File: src/server/game/Entities/Item/ItemTemplate.cpp (hunk 3 above)
Comment (Keader) [reply src/server/game/Entities/Item/ItemTemplate.cpp]: something like:

```cpp
bool ItemTemplate::CanBeDualWieldedWithTitansGrip() const
{
    SpellInfo const* spellInfo = sSpellMgr->AssertSpellInfo(46917); // Titan Grip
    return ((1 << SubClass) & spellInfo->EquippedItemSubClassMask) != 0;
}
```

**Comment by jackpoz** [comment]
Would anyone like to take over this PR ? Otherwise I will close it as abandoned

**Comment by svranesevic** [comment]
@jackpoz I'll wrap it up tomorrow

**Comment by jackpoz** [comment]
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.

**Pull request closed without merging**